<a href="https://colab.research.google.com/github/aymanattieh77/aymanattieh77/blob/main/carPricePredict.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**We want to predict car Prcie **


In [14]:
# import libaries
import pandas as pd # for data manipulation
import numpy as np # for linear algebra
import matplotlib.pyplot as plt # for data visualization
import seaborn as sns # for data visualization

In [15]:
# upload data
from google.colab import files
upload = files.upload()


Saving Car details v3.csv to Car details v3.csv


In [16]:
# read data 
df = pd.read_csv('Car details v3.csv')
df

name  year  selling_price  km_driven    fuel  \
0           Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1     Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2         Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3        Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4           Maruti Swift VXI BSIII  2007         130000     120000  Petrol   
...                            ...   ...            ...        ...     ...   
8123             Hyundai i20 Magna  2013         320000     110000  Petrol   
8124         Hyundai Verna CRDi SX  2007         135000     119000  Diesel   
8125        Maruti Swift Dzire ZDi  2009         382000     120000  Diesel   
8126               Tata Indigo CR4  2013         290000      25000  Diesel   
8127               Tata Indigo CR4  2013         290000      25000  Diesel   

     seller_type transmission                 owner     mileage   engine  \
0     Individual       Manual           First Owner   23.4 kmpl  1248 CC   
1     Individual       Manual          Second Owner  21.14 kmpl  1498 CC   
2     Individual       Manual           Third Owner   17.7 kmpl  1497 CC   
3     Individual       Manual           First Owner   23.0 kmpl  1396 CC   
4     Individual       Manual           First Owner   16.1 kmpl  1298 CC   
...          ...          ...                   ...         ...      ...   
8123  Individual       Manual           First Owner   18.5 kmpl  1197 CC   
8124  Individual       Manual  Fourth & Above Owner   16.8 kmpl  1493 CC   
8125  Individual       Manual           First Owner   19.3 kmpl  1248 CC   
8126  Individual       Manual           First Owner  23.57 kmpl  1396 CC   
8127  Individual       Manual           First Owner  23.57 kmpl  1396 CC   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2         78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...          ...                        ...    ...  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  
8126      70 bhp        140Nm@ 1800-3000rpm    5.0  
8127      70 bhp        140Nm@ 1800-3000rpm    5.0  

[8128 rows x 13 columns]

In [17]:
# types of columns
df.dtypes

name              object
year               int64
selling_price      int64
km_driven          int64
fuel              object
seller_type       object
transmission      object
owner             object
mileage           object
engine            object
max_power         object
torque            object
seats            float64
dtype: object

In [22]:
# check null values 
missing_values = df.isnull().sum()
print(missing_values)
total_missing_value = (missing_values.sum()/df.size)*100
print('Total missing values in dataset : ', total_missing_value)

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        215
torque           222
seats            221
dtype: int64
Total missing values in dataset :  1.0410357359176257


The cell above we have 222 null values in 5 columns, 
the Percentage of missing values is low so i will drop all rows that have null values *italicized text*

In [23]:
print('the shape before drop null values ', df.shape)
df.dropna(inplace=True)
print('the shape after drop null values ', df.shape)


the shape before drop null values  (8128, 13)
the shape after drop null values  (7906, 13)


In [59]:
from pandas.core.common import random_state
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import r2_score

def dataset_score(x_train, x_test, y_train, y_test):
     model1 = RandomForestRegressor(n_estimators=100, random_state=99)
     model1.fit(x_train, y_train)
     predict1 = model1.predict(x_test)
     model2 = DecisionTreeRegressor()
     model2.fit(x_train, y_train)
     predict2 = model2.predict(x_test)   
     model3 = LinearRegression()
     model3.fit(x_train, y_train)
     predict3 = model3.predict(x_test)
     
     result1 = {'RandomForestRegressor (MAE):':mean_absolute_error(y_test, predict1),
               'DecisionTreeRegressor (MAE):':mean_absolute_error(y_test, predict2),
               'LinearRegression (MAE):':mean_absolute_error(y_test, predict3)}
     result2 = {'RandomForestRegressor (r2):':r2_score(y_test, predict1),
               'DecisionTreeRegressor (r2):':r2_score(y_test, predict2),
               'LinearRegression (r2):':r2_score(y_test, predict3)}          
     return result1, result2           
     

In [32]:
# we need to covert categotical attributes to numeric attributes
# that will be done into method like Scaler encoder, but that will not preform well with columns have high number of unigue values greater than 15
# we need to drop columns with high cardinality
categorical_cols = df.select_dtypes(include='object')
high_cardinality = [cols for cols in categorical_cols if categorical_cols[cols].nunique()>10]



['name', 'mileage', 'engine', 'max_power', 'torque']

In [54]:
# we can covert engine to numeric
new_df = df.drop(high_cardinality, axis=1)


In [55]:
new_df.head()

year  selling_price  km_driven    fuel seller_type transmission  \
0  2014         450000     145500  Diesel  Individual       Manual   
1  2014         370000     120000  Diesel  Individual       Manual   
2  2006         158000     140000  Petrol  Individual       Manual   
3  2010         225000     127000  Diesel  Individual       Manual   
4  2007         130000     120000  Petrol  Individual       Manual   

          owner  seats  
0   First Owner    5.0  
1  Second Owner    5.0  
2   Third Owner    5.0  
3   First Owner    5.0  
4   First Owner    5.0

In [56]:
from sklearn.preprocessing import LabelEncoder
le= LabelEncoder()
for i in new_df.select_dtypes(include='object'):
  new_df[i] = le.fit_transform(new_df[i])

new_df



year  selling_price  km_driven  fuel  seller_type  transmission  owner  \
0     2014         450000     145500     1            1             1      0   
1     2014         370000     120000     1            1             1      2   
2     2006         158000     140000     3            1             1      4   
3     2010         225000     127000     1            1             1      0   
4     2007         130000     120000     3            1             1      0   
...    ...            ...        ...   ...          ...           ...    ...   
8123  2013         320000     110000     3            1             1      0   
8124  2007         135000     119000     1            1             1      1   
8125  2009         382000     120000     1            1             1      0   
8126  2013         290000      25000     1            1             1      0   
8127  2013         290000      25000     1            1             1      0   

      seats  
0       5.0  
1       5.0  
2       5.0  
3       5.0  
4       5.0  
...     ...  
8123    5.0  
8124    5.0  
8125    5.0  
8126    5.0  
8127    5.0  

[7906 rows x 8 columns]

In [57]:
# define featurs and the target varible
X = new_df.drop('selling_price', axis=1)
y = new_df['selling_price']


In [60]:
from sklearn.model_selection import train_test_split # for spliting the dataset into groups training and testing data

x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=100)

result1, result2 = dataset_score(x_train, x_test, y_train, y_test)
print(result1)
print()
print(result2)

{'RandomForestRegressor (MAE):': 161442.37047839866, 'DecisionTreeRegressor (MAE):': 173023.70545912287, 'LinearRegression (MAE):': 315542.77223534684}

{'RandomForestRegressor (r2):': 0.863882498368584, 'DecisionTreeRegressor (r2):': 0.8317591055901637, 'LinearRegression (r2):': 0.5027343183029207}
